OptimizelySDK/Odp/OdpManager.cs

```diff
@@ -0,0 +1,296 @@
+﻿/* 
+ * Copyright 2022 Optimizely
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using OptimizelySDK.ErrorHandler;
+using OptimizelySDK.Logger;
+using OptimizelySDK.Odp.Entity;
+using System;
+using System.Collections.Generic;
+
+namespace OptimizelySDK.Odp
+{
+    /// <summary>
+    /// Concrete implementation to orchestrate segment manager, event manager, and ODP config
+    /// </summary>
+    public class OdpManager : IOdpManager, IDisposable
+    {
+        /// <summary>
+        /// Denotes if ODP Manager is meant to be handling ODP communication
+        /// </summary>
+        private bool _enabled;
+
+        /// <summary>
+        /// Configuration used to communicate with ODP
+        /// </summary>
+        private volatile OdpConfig _odpConfig;
+
+        /// <summary>
+        /// Manager used to handle audience segment membership
+        /// </summary>
+        public IOdpSegmentManager SegmentManager { get; private set; }
+
+        /// <summary>
+        /// Manager used to send events to ODP 
+        /// </summary>
+        public IOdpEventManager EventManager { get; private set; }
+
+        /// <summary>
+        /// Logger used to record messages that occur within the ODP client
+        /// </summary>
+        private ILogger _logger;
+
+        /// <summary>
+        /// Update the settings being used for ODP configuration and reset/restart dependent processes
+        /// </summary>
+        /// <param name="apiKey">Public API key from ODP</param>
+        /// <param name="apiHost">Host portion of the URL of ODP</param>
+        /// <param name="segmentsToCheck">Audience segments to consider</param>
+        /// <returns>True if settings were update otherwise False</returns>
+        public bool UpdateSettings(string apiKey, string apiHost, List<string> segmentsToCheck)
+        {
+            var newConfig = new OdpConfig(apiKey, apiHost, segmentsToCheck);
+            if (_odpConfig.Equals(newConfig))
+            {
+                return false;
+            }
+
+            _odpConfig = newConfig;
```

Comment:
When ODPConfig is updated, the OdpConfig object itself is replaced in ODPManager, while copying fields in ODPEventManager and ODPSegmentManager. Can we make it consistent in either way?